File: skills/code-architect/references/code-examples/java/Memento.java
// Memento — capture and restore an object's state without breaking encapsulation.

import java.util.ArrayDeque;
import java.util.Deque;

class Editor {
    private String content = "";

    void type(String text) { content += text; }
    String content() { return content; }

    // Memento carries opaque state; only the originator can read it back.
    Snapshot save() { return new Snapshot(content); }
    void restore(Snapshot s) { this.content = s.state; }

    static final class Snapshot {
        private final String state;
        private Snapshot(String state) { this.state = state; }
    }
}

class Caretaker {
    private final Deque<Editor.Snapshot> stack = new ArrayDeque<>();
    void backup(Editor e) { stack.push(e.save()); }
    void undo(Editor e) { if (!stack.isEmpty()) e.restore(stack.pop()); }
}

public class Memento {
    public static void main(String[] args) {
        Editor editor = new Editor();
        Caretaker history = new Caretaker();

        editor.type("Hello");
        history.backup(editor);
        editor.type(", World");
        System.out.println("before undo: " + editor.content());

        history.undo(editor);
        System.out.println("after undo:  " + editor.content());
    }
}

File: skills/code-architect/references/code-examples/csharp/Memento.cs
// Memento — capture and restore an object's state without breaking encapsulation.
using System;
using System.Collections.Generic;

namespace Patterns.Memento
{
    // Opaque snapshot; the originator alone interprets its contents.
    sealed class Memento
    {
        public string State { get; }
        public Memento(string state) => State = state;
    }

    sealed class Originator
    {
        public string State { get; set; } = "";
        public Memento Save() => new(State);
        public void Restore(Memento memento) => State = memento.State;
    }

    // Caretaker stores mementos but never inspects their internals.
    sealed class Caretaker
    {
        private readonly Stack<Memento> _history = new();
        public void Backup(Originator o) => _history.Push(o.Save());
        public void Undo(Originator o)
        {
            if (_history.Count > 0) o.Restore(_history.Pop());
        }
    }

    static class Demo
    {
        public static void Run()
        {
            var originator = new Originator();
            var caretaker = new Caretaker();

            originator.State = "v1";
            caretaker.Backup(originator);
            originator.State = "v2";
            caretaker.Backup(originator);
            originator.State = "v3";
            Console.WriteLine($"Current: {originator.State}");

            caretaker.Undo(originator);
            Console.WriteLine($"After undo: {originator.State}");
            caretaker.Undo(originator);
            Console.WriteLine($"After undo: {originator.State}");
        }

        public static void Main() => Run();
    }
}
